Update required package versions

kllrak opened the pull request:
Supports Symbolics v3, ModelingToolkit v6, SymbolicUtils v0.17.

Tests pass locally, but I have not tried doctests. There a few TODOs related to knowing when to use `metadata`. I admit not fully understanding the metadata mechanism, what Symbolics or ModelingToolkit expect does not seem well documented yet. Also, for all the tests to pass, some type piracy was needed [here](https://github.com/qojulia/QuantumCumulants.jl/compare/master...kllrak:kllrak/upgrade?expand=1#diff-64c8deff57a563354497ae50b9f0b9fa59a65ae5cfb983898b7fc333791e02e8R2) (see also https://github.com/JuliaSymbolics/Symbolics.jl/pull/160, not sure why it is still open).

kllrak commented:
I ran the doctests and found that I had to modify a few more things in order for the `Position` and `Momentum` operator example in `docs/src/implementation.md` to work. Otherwise, the building the docs would fail by hanging indefinitely.

david-pl reviewed (commented):
Alright I finally got around to testing this. Overall it looks great, thanks again!

Some points I'd like to discuss:

- One question regarding the metadata: Why did you actually add metadata to all `QNumber` types? I don't really see a use for it, at least not yet. Do you have something in mind? Or is it just to make `QNumber`s behave more similar to `Symbolics`? Note that I also don't completely understand the metadata mechanisms, so really I'm just not sure here.
- While testing your PR locally I noticed that removing the introduced type piracy the tests still worked fine. I definitely remember there being a problem with this at some point, but I can't recall. All additional tests I ran also seem fine. Obviously I'm not a fan of that type piracy so can we remove it? Do you have an example that would fail if we do?
- Since MTK now determines variables and parameters according to context we should probably remove the `isparameter` method here https://github.com/qojulia/QuantumCumulants.jl/blob/430088cbdf9b9ba7061a872eaf746f496ed5a1d9/src/diffeq.jl#L57 and instead add the proper context to the metadata of parameters [like they do in MTK](https://github.com/SciML/ModelingToolkit.jl/blob/40b22eda6ff8c4f59913f9fe76520954bb750896/src/parameters.jl#L21)

kllrak commented on src/qnumber.jl:
```diff
@@ -1,3 +1,7 @@
+# TODO: Remove these once https://github.com/JuliaSymbolics/Symbolics.jl/pull/160 is merged
+Symbolics.toexpr(n::Complex{Symbolics.Num}, st) = :($Complex($(Symbolics.toexpr(real(n), st)), $(Symbolics.toexpr(imag(n), st))))
```
While I recall this being useful at some point, I agree that tests pass without it and I admit I don't recall why I added it. I will open a separate issue if I figure it out.

kllrak commented:
Hi David,

> Why did you actually add metadata to all QNumber types? I don't really see a use for it, at least not yet.

Metadata now seems to be expected for the ODESystem constructor to work properly. If you simply take the master branch of this package, update `[compat]` section of Project.toml to what is in this branch, and try to run tests, you'll first hit a failure in `test/test_diffeq.jl` because `ODESystem` constructor wants you to use the new `@named` macro. After fixing that, you'll find that tests error because they hit [this line](https://github.com/JuliaSymbolics/Symbolics.jl/blob/b7cae533e0a2aae8d7e227b2e0db4e25e6cb8d09/src/variable.jl#L375). Here's the error:

```julia
ArgumentError: Cannot find the parent of var"⟨a′*a⟩"(t).
  Stacktrace:
    [1] getparent(x::Term{Complex, Nothing}, val::Dict{Any, Any})
      @ Symbolics ~/.julia/packages/Symbolics/Cmx10/src/variable.jl:375
    [2] getparent
      @ ~/.julia/packages/Symbolics/Cmx10/src/variable.jl:367 [inlined]
    [3] _getname(x::Term{Complex, Nothing}, val::Dict{Any, Any})
      @ Symbolics ~/.julia/packages/Symbolics/Cmx10/src/variable.jl:356
    [4] getname (repeats 2 times)
      @ ~/.julia/packages/Symbolics/Cmx10/src/variable.jl:364 [inlined]
    [5] collect_var_to_name!(vars::Dict{Any, Any}, xs::Vector{Term{Complex, Nothing}})
      @ ModelingToolkit ~/.julia/packages/ModelingToolkit/P76eb/src/utils.jl:195
    [6] process_variables!(var_to_name::Dict{Any, Any}, defs::Dict{Any, Any}, vars::Vector{Term{Complex, Nothing}})
      @ ModelingToolkit ~/.julia/packages/ModelingToolkit/P76eb/src/utils.jl:174
    [7] ODESystem(deqs::Vector{Equation}, iv::Sym{Real, Nothing}, dvs::Vector{Any}, ps::OrderedCollections.OrderedSet{Any}; controls::Vector{Num}, observed::Vector{Num}, systems::Vector{ODESystem}, name::Symbol, default_u0::Dict{Any, Any}, default_p::Dict{Any, Any}, defaults::Dict{Any, Any}, connection_type::Nothing, preface::Nothing, checks::Bool)
      @ ModelingToolkit ~/.julia/packages/ModelingToolkit/P76eb/src/systems/diffeqs/odesystem.jl:135
    [8] ODESystem(eqs::Vector{Equation}, iv::Sym{Real, Nothing}; kwargs::Base.Iterators.Pairs{Symbol, Symbol, Tuple{Symbol}, NamedTuple{(:name,), Tuple{Symbol}}})
      @ ModelingToolkit ~/.julia/packages/ModelingToolkit/P76eb/src/systems/diffeqs/odesystem.jl:185
    [9] ODESystem(me::MeanfieldEquations, iv::Sym{Real, Nothing}; kwargs::Base.Iterators.Pairs{Symbol, Symbol, Tuple{Symbol}, NamedTuple{(:name,), Tuple{Symbol}}})
      @ QuantumCumulants /local/home/kllrak/.julia/dev/QuantumCumulants/src/diffeq.jl:61
   [10] macro expansion
      @ /local/home/kllrak/.julia/dev/QuantumCumulants/test/test_diffeq.jl:34 [inlined]
   [11] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [12] top-level scope
      @ /local/home/kllrak/.julia/dev/QuantumCumulants/test/test_diffeq.jl:9
   [13] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [14] top-level scope
      @ /local/home/kllrak/.julia/dev/QuantumCumulants/test/runtests.jl:33
   [15] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [16] top-level scope
      @ none:6
   [17] eval
      @ ./boot.jl:360 [inlined]
   [18] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:261
   [19] _start()
      @ Base ./client.jl:485
```

It may be that this could actually be considered unexpected behavior in Symbolics, in which case perhaps it would be better to fix the problem in Symbolics rather than add metadata to the types defined here. I do not understand the workings of Symbolics well enough to say whether that's a better route to go; it may be worth asking the Symbolics maintainers about this. Suffice it to say, I introduced metadata for compatibility rather than having a particular use case in mind.

Thanks for your other comments, which I have resolved. I also addressed your comment about `MTK.isparameter`.

david-pl commented:
Thanks for the update and for your answer!

> Metadata now seems to be expected for the ODESystem constructor to work properly

Just to clarify: everything concerning `QNumber` types is implemented in QuantumCumulants (algebra, simplifcation, etc.) and doesn't necessarily involve Symbolics. When averaging `QNumber` expressions we convert to symbolic number types that are handled by SymbolicUtils. Finally, when creating the `ODESystem` we convert the averages to variables in the form expected by Symbolics/MTK. So technically you can get away with adding metadata in the final step only and achieve the same thing. That's why I was a bit torn between whether we should introduce metadata at the level of `QNumber`s or not. For the most part we're just dragging around empty dictionaries which aren't really used. However, I now think that we might as well keep it. It may make it easier to have compatibility in the future and it might come in handy to be able to throw some random stuff into that metadata.

Interestingly performance appears to be taking a hit here, but I'm not sure why. It has nothing to do with the `QNumber` metadata (I tried without that) so it's probably some change that was done in the new Symbolics version. I'll need to discuss this with @ChristophHotter.

This PR is complete though in my opinion, so thanks again for all the work Andrew!
